src/index.ts

```diff
@@ -27,15 +30,27 @@ const createGoError = <E extends Error>(err: unknown): GoResultError<E> => {
   return fail(new Error('' + err));
 };
 
-export const go = async <T, E extends Error>(fn: Promise<T> | (() => Promise<T>)): Promise<GoResult<T, E>> => {
-  // We need try/catch because `fn` might throw sync errors as well
+export const go = async <T, E extends Error>(
+  fn: () => Promise<T>,
```

Comment:
Why did you get rid of the `Promise<T>` type? You can always wrap it later in the function but it's quite handy if you can pass the promise directly without wrapping it when using the library.